beancount_import/source/generic_importer_source.py

```diff
@@ -0,0 +1,188 @@
+        self.importer = importer
+        self.account = account
+
+        self._comparator = SimilarityComparator()
+
+        # get _FileMemo object for each file
+        files = [get_file(f) for f in
+                    filter(os.path.isfile,
+                 glob(os.path.join(directory, '**', '*'), recursive=True)
+                           )
+        ]
+        # filter the valid files for this importer
+        self.files = [f for f in files if self.importer.identify(f)]
+
+    @property
+    def name(self):
+        return self.importer.name()
+
+    def prepare(self, journal: 'JournalEditor', results: SourceResults) -> None:
+        results.add_account(self.account)
+        entries = defaultdict(list)
+        for f in self.files:
+            f_entries = self.importer.extract(f, existing_entries=journal.entries)
+            # collect  all entries in current statement, grouped by hash
+            hashed_entries = defaultdict(list)
+            for entry in f_entries:
+                hash_ = self._hash_entry(entry, frozenset(['filename','lineno']))
+                hashed_entries[hash_].append(entry)
+            # deduplicate across statements
+            for hash_ in hashed_entries:
+                # skip the existing entries from other statements. add remaining
+                n = len(entries[hash_])
+                entries[hash_].extend(hashed_entries[hash_][n:])
+
+        uncleared_entries = defaultdict(list)
+        for hash_ in entries:
+            # number of matching cleared entries in journal
+            n = len(similar_entries_in_journal(entries[hash_][0],
+                                               journal.entries,
+                                               self.comparator))
+            # If journal has n cleared entries for this hash, pick remaining
+            for entry in entries[hash_][n:]:
+                # add importer name as sorce description to source postings
+                self._add_description(entry)
+                # balance amount
+                self.balance_amounts(entry)
+                uncleared_entries[hash_].append(entry)
+
+        results.add_pending_entries(
+            [ImportResult(entry.date, [entry], None)
+                for entry in itertools.chain.from_iterable(uncleared_entries.values())
+            ]
+        )
+
+    def comparator(self, entry1, entry2):
+        """Returns if the two entries are similar and 2nd entry is cleared.
+        The first entry is from new_entries and 2nd is from journal
+        """
+        return self._comparator(entry1, entry2) \
+               and self.is_entry_cleared(entry2) \
+               and entry1.narration == entry2.postings[0].meta['source_desc']
+
+    def _add_description(self, entry: Transaction):
+        if not isinstance(entry, Transaction): return None
+        postings = entry.postings #type: ignore
+        to_mutate = []
+        for i, posting in enumerate(postings):
```

Comment:
In your tests the ingester always only generates a single posting, but I imagine in some cases it would generate multiple postings.  In that case, for compatibility with get_pending_and_invalid_entries, the source_desc property should only be added to one of the postings, e.g. the first one.

Reply:
Done by adding break!